File: backend/app/services/csv_parser.py
import pandas as pd
from datetime import datetime

def parse_portfolio_csv(file):
    try:
        df = pd.read_csv(file)
    except Exception as e:
        return None, [f"Pandas error: {str(e)}"]
    required_cols = ['Symbol', 'Quantity', 'Average Buy Price (INR)']
    errors = []
    for col in required_cols:
        if col not in df.columns:
            errors.append(f"Missing required column: {col}")

    if errors:
        return None, errors

    validated_indices = []
    row_errors = []
    for idx, row in df.iterrows():
        symbol = row.get('Symbol', '').strip()
        try:
            quantity = int(row.get('Quantity', 0))
            price = float(row.get('Average Buy Price (INR)', 0))
            purchase_date = str(row.get('Purchase Date (YYYY-MM-DD)', '')).strip()
        except Exception as e:
            row_errors.append(f"Row {idx+2}: Conversion error: {str(e)}")
            continue

        if (not symbol or quantity <= 0 or price <= 0 or
            (purchase_date and not valid_date(purchase_date))):
            row_errors.append(f"Row {idx+2}: Invalid data: {row.to_dict()}")
            continue
        validated_indices.append(idx)
    return df.loc[validated_indices].reset_index(drop=True), row_errors

def valid_date(date_str):
    try:
        datetime.strptime(date_str, "%Y-%m-%d")
        return True
    except:
        return False

File: backend/src/services/csvParser.js
const { parse } = require('csv-parse/sync');

/**
 * Parses a portfolio CSV buffer.
 * Mirrors parse_portfolio_csv() in csv_parser.py.
 *
 * Expected columns:
 *   Symbol | Quantity | Average Buy Price (INR) | Purchase Date (YYYY-MM-DD) [optional]
 *
 * Returns:
 *   { rows: Array<Object> | null, errors: string[] }
 */
function parsePortfolioCsv(buffer) {
  let records;
  try {
    records = parse(buffer, {
      columns:          true,    // first row = headers
      skip_empty_lines: true,
      trim:             true,
    });
  } catch (err) {
    return { rows: null, errors: [`CSV parse error: ${err.message}`] };
  }

  const REQUIRED = ['Symbol', 'Quantity', 'Average Buy Price (INR)'];
  const errors   = [];

  const headers = records.length > 0 ? Object.keys(records[0]) : [];
  for (const col of REQUIRED) {
    if (!headers.includes(col)) errors.push(`Missing required column: ${col}`);
  }
  if (errors.length) return { rows: null, errors };

  const rows = [];
  records.forEach((record, idx) => {
    const rowNum = idx + 2;           // row 1 = header
    const symbol = (record['Symbol'] || '').trim().toUpperCase();

    let quantity, averageBuyPrice, purchaseDate;
    try {
      quantity       = parseInt(record['Quantity'], 10);
      averageBuyPrice = parseFloat(record['Average Buy Price (INR)']);
      purchaseDate   = (record['Purchase Date (YYYY-MM-DD)'] || '').trim() || null;
    } catch (e) {
      errors.push(`Row ${rowNum}: Conversion error: ${e.message}`);
      return;
    }

    if (!symbol || isNaN(quantity) || quantity <= 0 ||
        isNaN(averageBuyPrice) || averageBuyPrice <= 0) {
      errors.push(`Row ${rowNum}: Invalid data: ${JSON.stringify(record)}`);
      return;
    }

    if (purchaseDate && !isValidDate(purchaseDate)) {
      errors.push(`Row ${rowNum}: Invalid date format (expected YYYY-MM-DD): ${purchaseDate}`);
      return;
    }

    rows.push({ symbol, quantity, averageBuyPrice, purchaseDate });
  });

  return { rows, errors };
}

function isValidDate(str) {
  return /^\d{4}-\d{2}-\d{2}$/.test(str) && !isNaN(Date.parse(str));
}

module.exports = { parsePortfolioCsv };